features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/ResourceResolver.kt

```diff
@@ -0,0 +1,284 @@
+            shouldCacheBitmap = shouldCacheBitmap,
+            bitmap = bitmap,
+            resourceId = resourceId,
+            drawable = drawable
+        )
+
+        resolveResourceCallback.onResolved(resourceId, byteArray)
+    }
+
+    private fun cacheIfNecessary(
+        shouldCacheBitmap: Boolean,
+        bitmap: Bitmap,
+        resourceId: String,
+        drawable: Drawable
+    ) {
+        if (shouldCacheBitmap) {
+            bitmapCachesManager.putInBitmapPool(bitmap)
+        }
+
+        bitmapCachesManager.putInResourceCache(drawable, resourceId)
+    }
+
+    @WorkerThread
+    private fun tryToDrawNewBitmap(
+        resources: Resources,
+        drawable: Drawable,
+        drawableWidth: Int,
+        drawableHeight: Int,
+        displayMetrics: DisplayMetrics,
+        resolveResourceCallback: ResolveResourceCallback
+    ) {
+        drawableUtils.createBitmapOfApproxSizeFromDrawable(
+            resources = resources,
+            drawable = drawable,
+            drawableWidth = drawableWidth,
+            drawableHeight = drawableHeight,
+            displayMetrics = displayMetrics,
+            bitmapCreationCallback = object : BitmapCreationCallback {
+                override fun onReady(bitmap: Bitmap) {
+                    val byteArray = webPImageCompression.compressBitmap(bitmap)
+
+                    @Suppress("ThreadSafety") // this runs inside an executor
+                    resolveResourceHash(
+                        drawable = drawable,
+                        bitmap = bitmap,
+                        byteArray = byteArray,
+                        shouldCacheBitmap = true,
+                        resolveResourceCallback = resolveResourceCallback
+                    )
+                }
+
+                override fun onFailure() {
+                    resolveResourceCallback.onFailed()
+                }
+            }
+        )
+    }
+
+    @WorkerThread
+    @Suppress("ReturnCount")
+    private fun tryToGetBitmapFromBitmapDrawable(
+        drawable: BitmapDrawable,
+        bitmapFromDrawable: Bitmap,
+        resolveResourceCallback: ResolveResourceCallback
+    ): Bitmap? {
+        drawableUtils.createScaledBitmap(bitmapFromDrawable)?.let { scaledBitmap ->
+            val byteArray = webPImageCompression.compressBitmap(scaledBitmap)
+
+            // failed to get byteArray potentially because the bitmap was recycled before imageCompression
+            // we'll now failover to attempting to create a new bitmap from the drawable
+            if (byteArray.isEmpty() && scaledBitmap.isRecycled) {
+                return null
+            }
+
+            /**
+             * Check whether the scaled bitmap is the same as the original.
+             * Since Bitmap.createScaledBitmap will return the original bitmap if the
+             * requested dimensions match the dimensions of the original
+             */
+            val shouldCacheBitmap = scaledBitmap != drawable.bitmap
```

Comment:
> Since Bitmap.createScaledBitmap will return the original bitmap if the requested dimensions match the dimensions of the original

Does it mean we should/may use `!==` (reference comparison) here instead? How equality operator is defined for bitmaps, will it go through the bitmap content?

Reply:
Using reference comparison seems to return false negatives. I'm changing this to compare width and height directly

Reply:
well, it seems the current one `!=` is doing reference comparison already, because I don't see any `equals` override in the source code https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/graphics/java/android/graphics/Bitmap.java.

Reply:
yes, this was potentially causing caching where it shouldn't have occurred. I think it's best to use the width/height. That's what we're actually checking here - has createScaledBitmap actually created a resized bitmap